src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,319 @@
+//  Next we need to translate delegations which should pass beliefs from one
+//  principal to another. For example, if we have
+//      A says B canSay fact(x) :- fact1(x)
+//  we do the usual translation resulting in
+//      (1) A says B canSay fact(x) :- A says fact1(x)
+//  but we also add the additional rule
+//      (2) A says fact(x) :- x says fact(x), A says x canSay fact(x)
+//  where x is a fresh variable.
+//  The result is that rule (1) gives the condition for delegation
+//  and rule (2) passes beliefs from B to A when the condition from (1) is met.
+//
+// Finally we need to translate canActAs which should pass properties from
+// one principal to another:
+// Assertions of either the form
+//      A says B canActAs C :- ...
+// or
+//      A says B <anyPredicate(...)> :-
+// are claims by A that B has a property (where canActAs is just a special
+// case). Syntactically these are both treated as "A says B <verbphrase>" with
+// different instances of <verbphrase>. Whenever some other principal can act
+// as B, we want to pass these properties from B to that other principal.
+// (For the special case where this property is B canActAs C, this essentially
+// makes canActAs transitive). In either case, we add an additional rule
+//    A says x verbphrase <- A says x canActAs B, A says B verbphrase
+// which will add "verbphrase" as a property of D whenever we can prove
+// the assertion "A says D canActAsB".
+//
+// ## Translation of Queries
+// Neither Souffle nor SecPal support queries, but this language does.
+// Queries check of a single unconditional assertion is true. A query of
+// the form:
+// ```Datalog
+//      Q_NAME = query ASSERTION(<ARGS>) ?
+// ```
+//
+// is translated into an assertion with just one argument, and this
+// assertion is made an output, so we can simply check the CSV that
+// souffle emits as in:
+// ```Datalog
+//      Q_NAME("dummy_var") :- grounded("dummy_var"), ASSERTION(<ARGS>).
+//      grounded("dummy_var").
+//      .output Q_NAME
+// ```
+
+class LoweringToDatalogPass {
+ public:
+  static DLIRProgram Lower(const Program& auth_logic_program) {
+    return LoweringToDatalogPass().ProgToDLIR(auth_logic_program);
+  }
+
+ private:
+  LoweringToDatalogPass() : fresh_var_count_(0) {}
+
+  // This returns a fresh variable name. Fresh variables are introduced in a
+  // few places during the translation.
+  std::string FreshVar() {
+    CHECK(fresh_var_count_ < std::numeric_limits<uint64_t>::max());
+    return "x__" + std::to_string(++fresh_var_count_);
+  }
+
+  // This function takes a `predicate` and creates a new predicate as follows:
+  //   - Prepend the `modifier` to the name of the given `predicate`.
+  //   - Prepend the given `args` to the original list of arguments of
+  //   `predicate`.
+  //
+  // This is used in a few places in the translation, for example, to translate
+  // "X says blah(args)" into "says_blah(X, args)".
+  Predicate PushOntoPredicate(absl::string_view modifier,
```

Comment:
Actually, all functions from here on could move to the cc file.

Reply:
For some reason I thought you just meant all the functions after these internal helpers. I moved this one and the one below it over now.